### ISSUE TITLE
UnsupportedOperationException when reading from the queue

### ISSUE THREAD

**Issue by javasoze** [OpenHFT/Chronicle-Queue#628]
CQ version: 5.17.25

```
! java.lang.UnsupportedOperationException: FIELD_207
! at net.openhft.chronicle.wire.BinaryWire$BinaryValueIn.cantRead(BinaryWire.java:3498)
! at net.openhft.chronicle.wire.BinaryWire$BinaryValueIn.int64array(BinaryWire.java:2823)
! at net.openhft.chronicle.queue.impl.single.SCQIndexing.array(SCQIndexing.java:173)
! at net.openhft.chronicle.queue.impl.single.SCQIndexing.arrayForAddress(SCQIndexing.java:162)
! at net.openhft.chronicle.queue.impl.single.SCQIndexing.moveToIndex0(SCQIndexing.java:269)
! at net.openhft.chronicle.queue.impl.single.SCQIndexing.moveToIndex(SCQIndexing.java:231)
! at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueStore.moveToIndexForRead(SingleChronicleQueueStore.java:219)
! at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.moveToIndexResult(SingleChronicleQueueExcerpts.java:1422)
! at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.moveToIndexInternal(SingleChronicleQueueExcerpts.java:1468)
! at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.nextEndOfCycle(SingleChronicleQueueExcerpts.java:1170)
! at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.endOfCycle(SingleChronicleQueueExcerpts.java:1152)
! at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.next0(SingleChronicleQueueExcerpts.java:1124)
! at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.readingDocument(SingleChronicleQueueExcerpts.java:1050)
! at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.readingDocument(SingleChronicleQueueExcerpts.java:980)
! at net.openhft.chronicle.wire.MarshallableIn.readBytes(MarshallableIn.java:64)
```

**Comment by dpisklov** [comment]
Any reproduction steps?

**Comment by javasoze** [comment]
@dpisklov thanks for responding so quickly!

We ran into this in production randomly, so we don't know how to reproduce this ourselves.

But we have been often invalid entries errors where we needed to implement in our code to skip these entries. Not sure if that is related...

**Comment by dpisklov** [comment]
Can you at least share the code you use to read/write to/from the queue?

**Comment by javasoze** [comment]
On the producing(write) side, here is the code snippet:
```
// setting it up
var vqueue = SingleChronicleQueueBuilder
            .binary(firehose.getConfig().directory)
            .rollCycle(RollCycles.MINUTELY)
            .strongAppenders(true)
            .build();
ExcerptAppender appender = queue.acquireAppender();

// writing to cq, this is in a method basically listen to events and writes to the queue
try {
            appender.writeBytes(b -> {
                b.writeLong(builder.getTimestamp());
                b.write(buffer.array(), buffer.position(), buffer.limit() - buffer.position());
            });
        } catch (Exception e) {
            LOGGER.error("Exception thrown while writing to chronicle queue. Restarting the process", e);
            System.exit(-1);
        }
```

On the reading side, see the following code snippet:

```
public class ChronicleQueueFirehose implements Firehose {
    private static final Logger LOGGER = LoggerFactory.getLogger(ChronicleQueueFirehose.class);

    private static final Histogram CHRONICLE_QUEUE_READ_HISTOGRAM =
        Histogram.build()
            // 100KB bucket * 10
            .linearBuckets(0, 1024 * 100, 10)
            .name("dashbase_index_chronicle_queue_read")
            .help("Histogram of event size read from chronicle queue")
            .register();

    private static final Gauge NUM_FILES =
        Gauge.build("dashbase_index_chronicle_queue_num_files", "number of chronicle queue files")
            .register();

    private static final Gauge NUM_ENTRIES =
        Gauge.build("dashbase_index_chronicle_queue_num_entries", "number of entries in chronicle queue")
            .register();

    private static final Gauge SKIP_TO_END =
        Gauge.build("dashbase_index_chronicle_queue_skip_to_end",
            "number of times to skip to the end of the queue due to data corruption")
            .register();

    private static final Gauge HEALTH = Gauge.build(
        "dashbase_index_chronicle_queue_health",
        "status of chronicle queue health check").register();


    private static final ThreadLocal<ExcerptTailer> THREAD_LOCAL_TAILER = new ThreadLocal<>();

    private final ByteBuffer buffer = ByteBuffer.allocate(1024 * 1024);
    private final Config config;
    private final Map<File, Integer> releasedFiles = new ConcurrentHashMap<>();
    private final SingleChronicleQueue queue;
    private final HealthCheck healthCheck;

    // use a different queue to get the last index
    // in order to avoid releasing rolled files.
    private final SingleChronicleQueue queueForTail;

    private ExcerptTailer tailer;

    private volatile long indexToResume = -1L;
    private volatile long lastIndex = -1;
    private volatile int cycleOfLastIndex = -1;
    private volatile long lastTimestamp = -1;
    private volatile boolean isShutdown = false;

    @JsonCreator
    public ChronicleQueueFirehose(Config config) {
        this.config = config;
        this.queue = SingleChronicleQueueBuilder
            .binary(config.directory)
            .rollCycle(RollCycles.MINUTELY)
            .readOnly(true)
            .storeFileListener((cycle, file) -> {
                if (cycle < cycleOfLastIndex) {
                    file.delete();
                    LOGGER.info(file + " is deleted.");
                } else if (releasedFiles.putIfAbsent(file, cycle) == null) {
                    LOGGER.info(file + " is released but not deleted yet");
                    LOGGER.debug("cycle of the file is {}. cycle of last segment is {}", cycle, cycleOfLastIndex);
                }
            })
            .build();

        this.queueForTail = SingleChronicleQueueBuilder
            .binary(config.directory)
            .rollCycle(RollCycles.MINUTELY)
            .readOnly(true)
            .build();

        this.healthCheck = new HealthCheck();

        NUM_FILES.setChild(new Gauge.Child() {
            @Override
            public double get() {
                return getNumberOfCQFiles();
            }
        });

        NUM_ENTRIES.setChild(new Gauge.Child() {
            @Override
            public double get() {
                return getNumberOfEntries();
            }
        });
    }

    @Override
    public void close() throws IOException {
        if (isShutdown) {
            return;
        }
        LOGGER.info("Closing chronicle queue");
        isShutdown = true;
        queue.close();
        queueForTail.close();
    }

    @Override
    public boolean seekToOffset(String offset) {
        // This method is called from main thread, not from FirehoseIngester, so we cannot create and manipulate
        // tailer (which is not thread safe) here. So this method just records the offset into indexToResume,
        // which will be used in initTailer() later.

        if (OFFSET_EARLIEST.equalsIgnoreCase(offset)) {
            this.indexToResume = -1;
            return true;
        } else if (OFFSET_LATEST.equalsIgnoreCase(offset)) {
            this.indexToResume = Long.MAX_VALUE;
            return true;
        }

        try {
            this.indexToResume = Long.parseLong(offset);
            return true;
        } catch (NumberFormatException e) {
            LOGGER.error("Invalid offset: " + offset);
        }
        return false;
    }

    @Override
    public String getOffset() {
        return String.valueOf(lastIndex);
    }

    @Override
    public void release(String offset) {
        // delete the released files if their cycles are less than the cycle of the current index.
        long index = Long.parseLong(offset);
        if (index < 0) {
            return;
        }
        cycleOfLastIndex = queue.rollCycle().toCycle(index);
        synchronized (releasedFiles) {
            releasedFiles.entrySet().removeIf((e) -> {
                var file = e.getKey();
                var cycle = e.getValue();
                if (cycle < cycleOfLastIndex) {
                    file.delete();
                    LOGGER.info(file + " is deleted.");
                    return true;
                }
                return false;
            });
        }
    }

    @Override
    public long getTimestampInMillis() {
        return lastTimestamp;
    }

    private void initTailer() {
        if (tailer == null) {
            tailer = queue.createTailer();

            if (indexToResume < 0) {
                LOGGER.info("Start reading from the beginning.");
                return;
            } else if (indexToResume == Long.MAX_VALUE) {
                LOGGER.info("Start reading from the latest.");
                tailer = tailer.toEnd();
                lastIndex = getEndIndex();
                return;
            }

            try {
                LOGGER.info("Moving index to " + indexToResume);
                if (tailer.moveToIndex(indexToResume)) {
                    lastIndex = indexToResume;
                    release(String.valueOf(indexToResume));
                    // the current entry is already indexed, so need to skip.
                    try (var dc = tailer.readingDocument()) {}
                } else {
                    LOGGER.warn("Not able to resume from index {}. starting from the beginning of the queue", indexToResume);
                    tailer.toStart();
                }
            } catch (Exception e) {
                LOGGER.warn("Failed to move index", e);
                try {
                    tailer.toEnd();
                    long indexOfStart = tailer.index();
                    LOGGER.warn("Failed to resume from index {}. Read from the latest {}", indexToResume, indexOfStart);
                } catch (Exception e2) {
                    LOGGER.error("Failed to resume from the latest. Will delete the files and restart.");
                    Arrays.stream(queue.file().listFiles()).forEach(file -> file.delete());
                    System.exit(-1);
                }
            }
        }
    }

    @Override
    public ByteBuffer readToByteBuffer() {
        initTailer();
        try {
            if (!tailer.readBytes(b -> {
                lastTimestamp = b.readLong();
                buffer.clear();
                b.read(buffer);
                buffer.flip();
                CHRONICLE_QUEUE_READ_HISTOGRAM.observe(buffer.limit() - buffer.position() + 8);
                lastIndex = tailer.index();
            })) {
                return null;
            }

            return buffer;
        } catch (Error e) {
            LOGGER.error("Data in the queue was corrupted. Will skip to the end of the queue", e);
            SKIP_TO_END.inc();
            tailer.toEnd();
            lastIndex = tailer.index();
            return null;
        }
    }

    @Override
    public byte[] next() throws IOException {
        ByteBuffer buffer = readToByteBuffer();
        if (buffer == null) {
            return null;
        }
        byte[] b = new byte[buffer.remaining()];
        buffer.get(b);
        return b;
    }

    private long getEndIndex() {
        var tailer = THREAD_LOCAL_TAILER.get();
        if (tailer == null) {
            tailer = queueForTail.createTailer();
            THREAD_LOCAL_TAILER.set(tailer);
        }
        return tailer.toEnd().index();
    }

    public int getNumberOfCQFiles() {
        // there is always one metadata file
        return new File(config.directory).list().length - 1;
    }

    public long getNumberOfEntries() {
        if (lastIndex < 0) {
            return 0;
        }
        long endIndex = getEndIndex();
        return endIndex == 0 ? 0 : Math.max(queueForTail.countExcerpts(lastIndex, getEndIndex()) - 1, 0);
    }

    public Config getConfig() {
        return config;
    }

    public static class Config {
        // directory to store Chronicle Queue files
        public String directory = "/dashbase/queue";

        // max throughput to write to Chronicle Queue in MB/s
        // writer will be rate limited if it's trying to write faster than this.
        // Setting this to 0 disables the rate limiting.
        public int maxWriteThroughputMBs = 20;
        
        // If Chronicle Queue contains more entries than this threshold,
        // IndexResource will add additional latency in returning responses
        // so that the log sender will back off.
        // This parameter can be turned to reduce a change where entries are
        // added faster than ChronicleQueueFirehose can consume.
        public int numEntriesToBackOff = 500000;

        // IndexResource will back off up to this value.
        public int maxBackOffSecond = 5;

        // If Chronicle Queue contains more entries than this threshold,
        // IndexResource will stop receiving further requests
        // by changing the health check status to UNHEALTHY.
        public int maxNumEntries = 1000000;

        // If Chronicle Queue is not empty but the number of entries in the queue
        // doesn't change for {restartAfterStallInSec} seconds, Table will restart itself.
        // This could happen if Chronicle Queue is writing faster than
        // Disk throughput, for example.
        public int restartAfterStallInSec = 180;
    }

    public HealthCheck getHealthCheck() {
        return healthCheck;
    }

    public class HealthCheck implements Runnable {
        private final Optional<RateLimiter> rateLimiter;

        private volatile int backOffSec = 0;
        private volatile long previousNumEntries = Long.MIN_VALUE;
        private volatile int count = 0;
        private volatile boolean isRunning = true;

        public HealthCheck() {
            HEALTH.setChild(new Gauge.Child() {
                @Override
                public double get() {
                    return isHealthy() ? 0 : 1;
                }
            });

            this.rateLimiter = Optional.of(config.maxWriteThroughputMBs)
                .filter(r -> r > 0).map(t -> RateLimiter.create(t * 1024 * 1024));
        }

        @Override
        public void run() {
            if (!isRunning) {
                return;
            }

            long numEntries = getNumberOfEntries();

            // calculate back off period
            if (numEntries < config.numEntriesToBackOff) {
                backOffSec = 0;
            } else if (numEntries > config.maxNumEntries) {
                backOffSec = -1;
            } else {
                backOffSec = (int) (config.maxBackOffSecond * 1000.0
                    * (numEntries - config.numEntriesToBackOff) / (config.maxNumEntries - config.numEntriesToBackOff));
            }

            // check if CQ is stalled
            if (numEntries > 0 && numEntries == previousNumEntries) {
                count++;
            } else {
                count = 0;
            }

            if (count > config.restartAfterStallInSec) {
                LOGGER.error("Chronicle Queue seems to be stalled. Restarting the process.");
                System.exit(-1);
            }
            previousNumEntries = numEntries;
        }

        public boolean isHealthy() {
            return isRunning && getNumberOfEntries() <= config.maxNumEntries;
        }

        public void rateLimit(int bytes) {
            rateLimiter.ifPresent(r -> r.acquire(bytes));
        }

        /**
         * Returns true if chronicle queue is ready (= not full).
         * If it's partially full, i.e., if # of entries > numEntriesToBackOff but < maxNumEntries,
         * then it also waits for back off.
         */
        public boolean isReady() {
            if (!isRunning) {
                return false;
            }
            if (backOffSec > 0) {
                try {
                    Thread.sleep(backOffSec);
                } catch (InterruptedException e) {}
            }
            return backOffSec >= 0;
        }

        public void close() {
            isRunning = false;
        }
    }
}
```

**Comment by javasoze** [comment]
We are writing and reading at a high rate, e.g. > 10MB/s. Each even is around 1KB.

We are deployed in Kubernetes, running in the cloud and using the cloud vendor's persistent volume.

**Comment by dpisklov** [comment]
@javasoze this indeed can be caused by the issue you described in #629 as that will  most likely corrupt queue file. I'll add more comments there and close this issue for now - if after fixing #629 you still see this one reoccurring please reopen.

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.29](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.29), [BOM-2.17.472](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.472)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.29](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.29), [BOM-2.17.496](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.496)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.29](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.29), [BOM-2.17.537](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.537)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.29](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.29), [BOM-2.17.541](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.541)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.29](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.29), [BOM-2.17.544](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.544)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.29](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.29), [BOM-2.17.546](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.546)
